evm_loader/program/src/executor/state.rs

```diff
@@ -285,17 +285,39 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
     }
 
     async fn code_hash(&self, from_address: Address, chain_id: u64) -> Result<[u8; 32]> {
-        use solana_program::keccak::hash;
-
-        for action in &self.actions {
-            if let Action::EvmSetCode { address, code, .. } = action {
-                if &from_address == address {
-                    return Ok(hash(code).to_bytes());
-                }
-            }
+        // https://eips.ethereum.org/EIPS/eip-1052
+        // https://eips.ethereum.org/EIPS/eip-161
+
+        #[maybe_async]
+        async fn data_account_exists<B: AccountStorage>(
+            state: &ExecutorState<'_, B>,
+            address: Address,
+            chain_id: u64,
+        ) -> Result<bool> {
+            Ok(state.nonce(address, chain_id).await? > 0
+                || state.balance(address, chain_id).await? > 0)
         }
 
-        Ok(self.backend.code_hash(from_address, chain_id).await)
```

Comment:
This was the only usage of `AccountStorage::code_hash` method, so it can now be removed from `AccountStorage` trait, with both its duplicated implementations from `EmulatorAccountStorage` and `ProgramAccountStorage`.

`AccountStorage` trait:
```rust
    async fn code_hash(&self, address: Address, chain_id: u64) -> [u8; 32];
```

`EmulatorAccountStorage` impl:
```rust
    async fn code_hash(&self, address: Address, chain_id: u64) -> [u8; 32] {
        use solana_sdk::keccak::hash;

        info!("code_hash {address} {chain_id}");

        let code = self.code(address).await.to_vec();
        if !code.is_empty() {
            return hash(&code).to_bytes();
        }

        // https://eips.ethereum.org/EIPS/eip-1052
        // https://eips.ethereum.org/EIPS/eip-161
        if (self.balance(address, chain_id).await == 0)
            && (self.nonce(address, chain_id).await == 0)
        {
            return <[u8; 32]>::default();
        }

        hash(&[]).to_bytes()
    }
```

`ProgramAccountStorage` impl:
```rust
    fn code_hash(&self, address: Address, chain_id: u64) -> [u8; 32] {
        use solana_program::keccak;

        if let Ok(contract) = self.contract_account(address) {
            keccak::hash(&contract.code()).to_bytes()
        } else {
            // https://eips.ethereum.org/EIPS/eip-1052
            // https://eips.ethereum.org/EIPS/eip-161
            if let Ok(account) = self.balance_account(address, chain_id) {
                if account.nonce() > 0 || account.balance() > 0 {
                    // account without code
                    keccak::hash(&[]).to_bytes()
                } else {
                    // non-existent account
                    <[u8; 32]>::default()
                }
            } else {
                // non-existent account
                <[u8; 32]>::default()
            }
        }
    }
```

The remaining `Database::code_hash` method implementation from `ExecutorState::code_bash` is equivalent to the two implementations above.

`Database` trait:
```rust
    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]>;
```

`ExecutorState` impl:
```rust
        // https://eips.ethereum.org/EIPS/eip-1052
        // https://eips.ethereum.org/EIPS/eip-161

        let code = self.code(address).await?;

        let bytes_to_hash: Option<&[u8]> = if !code.is_empty() {
            // A program account exists at the address.
            Some(&code)
        } else if self.nonce(address, chain_id).await? > 0
            || self.balance(address, chain_id).await? > 0
        {
            // A data account exists at the address.
            Some(&[])
        } else {
            // No account exists at the address.
            None
        };

        Ok(bytes_to_hash
            .map(|bytes| solana_program::keccak::hash(bytes).to_bytes())
            .unwrap_or_default())
```

In general, `ExecutorState` uses a type implementing `AccountStorage` inside it (`ProgramAccountStorage` or `EmulatorAccountStorage`) and the remaining implementation delegates calls to the `AccountStorage` type as needed.

Reply:
Merged your suggestion PR.